A selection of cross-compiling fixes

bgamari opened the pull request:
###### Motivation for this change

Allow a variety of core packages to be cross-compiled. This is a batch of patches pulled from my larger batch in #30882. Many of the rest were kindly merged by @dtzWill. The few that remain after this will require a bit more work.

 I've been using these to cross-compile ARMv7 NixOS images for several months now.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bgamari committed:
curl: Enable cross-compilation

By fixing toolchain paths

bgamari committed:
runit: Enable cross-compilation

bgamari committed:
libnl: Enable parallel building

This is a long enough derivation to make this worthwhile.

bgamari committed:
tcp-wrappers: Enable cross-compilation

This package, like so many others, assumes that ar is unprefixed.

bgamari committed:
net-tools: Enable cross-compilation

bgamari committed:
mmc-utils: Enable cross-compilation

bgamari committed:
avahi: Enable cross-compilation

bgamari committed:
ldns: Enable cross-compilation

bgamari committed:
rng-tools: Enable cross-compilation

bgamari commented:
Sadly I can't build this with `nox-review`; it fails with `Too many heap sections: Increase MAXHINCR or MAX_HEAP_SECTS`, presumably due to the very large size of the build.

matthewbauer commented on pkgs/os-specific/linux/libnl/default.nix:
```diff
@@ -19,6 +19,8 @@ stdenv.mkDerivation {
       sha256 = "0dd7xxikib201i99k2if066hh7gwf2i4ffckrjplq6lr206jn00r";
     });
 
+  enableParallelBuilding = true;
```
Not that this hurts anything - but I'm kind of amazed this would fix cross compiling. Usually parallel building breaks stuff. This is the first time I've heard of it fixing anything.

dtzWill replied:
It helps all the time when cross-compiling*!

* not fixing errors but making the rebuild/rebuild/rebuild go a bit faster and reduces the toll on your sanity!

bgamari replied:
> not fixing errors but making the rebuild/rebuild/rebuild go a bit faster and reduces the toll on your sanity!

Couldn't have said it better myself.

dtzWill commented on pkgs/os-specific/linux/tcp-wrappers/default.nix:
```diff
@@ -20,6 +20,7 @@ in stdenv.mkDerivation rec {
   prePatch = ''
     tar -xaf $debian
     patches="$(cat debian/patches/series | sed 's,^,debian/patches/,') $patches"
+    substituteInPlace Makefile --replace "AR	= ar" "AR = ${stdenv.cc.targetPrefix}ar"
```
FWIW it might be best (if you disagree nevermind it's basically the same) if instead of patching the Makefile we just add `"AR:=${stdenv.cc.targetPrefix}ar"`,  or `"AR:=$(AR)"` actually, to makeFlags?

bgamari replied:
I have absolutely no opinion. I will happily do whatever you all deem best.

bgamari replied:
Wow, I am slightly embarrassed to admit that I had no idea that `make` variables given on the command-line override definitions in the Makefile. Yes, overriding seems like a significantly better option.

Ericson2314 replied:
Or is this `:=` overriding `=`?

Ericson2314 replied:
To be clear, I didn't know this either, and I approve of @dtzWill's approach.

Ericson2314 replied:
`"AR:=$(AR)"` is especially nifty.

dtzWill commented on pkgs/tools/security/rng-tools/default.nix:
```diff
@@ -9,6 +9,12 @@ stdenv.mkDerivation rec {
+  # For cross-compilation
+  configureFlags = [ "AR=${stdenv.cc.bintools.targetPrefix}ar" ];
+  postPatch = ''
+    substituteInPlace Makefile.in --replace "AR = ar" ""
```
Also, since AR in the environment is already target-prefixed, are the configureFlags needed?

Ericson2314 replied:
`"AR:=$(AR)"` per the above.

bgamari commented on pkgs/os-specific/linux/nfs-utils/default.nix:
```diff
@@ -37,7 +39,10 @@ in stdenv.mkDerivation rec {
     ]
-    ++ lib.optional (stdenv ? glibc) "--with-rpcgen=${stdenv.glibc.bin}/bin/rpcgen";
+    ++ lib.optional (stdenv ? glibc) "--with-rpcgen=${stdenv.glibc.bin}/bin/rpcgen"
+    ++ stdenv.lib.optionals (stdenv.hostPlatform != stdenv.buildPlatform) [
```
Huh, indeed this doesn't appear to be necessary any more yet I thought I had already kicked out the patches that proved to be redundant.

bgamari replied:
I'll drop this patch.

bgamari committed:
lib.systems.gnu: Accept gnueabi as a gnu platform

bgamari commented on pkgs/tools/system/runit/default.nix:
```diff
@@ -34,8 +38,10 @@ stdenv.mkDerivation rec {
-    echo cc > conf-ld
+    echo ${stdenv.cc.targetPrefix}cc > conf-cc
+    echo ${stdenv.cc.targetPrefix}cc > conf-ld
+    export AR=${stdenv.cc.targetPrefix}ar
```
Hmm, I suppose this also isn't needed anymore.

Ericson2314 replied:
Oh good point

dtzWill commented:
All good, I surely know what it's like to sit on a huge stack of patches and test them over and over while splitting them out.  And always there's a detail or two I overlooked! That I swear wasn't like that or was needed a month ago or in a cross-env ... hehe.

Anyway, thanks for pushing these through and look forward to next iteration! :)

bgamari committed:
glib: Enable cross-compilation

glib needs gettext for the build machine. Since gettext is also a
propagatedBuildInput I'm assuming that it's also used at runtime.

bgamari commented:
I two more low-risk fixes to this branch: `glib` and `avahi`.

Ericson2314 commented:
@bgamari looks good except you actually added c0e0bc8 and 24001705e6bdaf64d8e68b26823dc379acfaa246?

bgamari commented:
Fixed.

Ericson2314 commented:
Thanks @bgamari

dtzWill commented on lib/systems/for-meta.nix:
```diff
@@ -19,7 +19,11 @@ in rec {
+    { kernel = parse.kernels.linux; abi = parse.abis.gnu; }
+    { kernel = parse.kernels.linux; abi = parse.abis.gnueabi; }
+    { kernel = parse.kernels.linux; abi = parse.abis.gnueabihf; }
+  ];
```
Just curious: why is this not just matching on ABI's? It seems unintuitive for "this package runs on GNU" to mean "gnu/linux" (and as a recent PR addressed seems to have been abused to mean "linux" in many packages).

But would it be "bad" to not require the kernel be linux?

bgamari replied:
A good question. I was looking to minimize the potential risk incurred by my patch, but I'll admit that this definition is rather fishy.

dtzWill replied:
Haha, understood re:conservative changes.  I do the same thing all the time.  cc @Ericson2314 -- do you know the answer?

Not a big deal ... :)

Ericson2314 replied:
I'm fine either way. I made some changes upstream to GNU Config that are leading me to reconsider some of this stuff a bit. E.g. maybe ABI should be a per-kernel namespace.

 - http://lists.gnu.org/archive/html/config-patches/2018-04/threads.html
 - http://lists.gnu.org/archive/html/config-patches/2018-05/threads.html

bgamari commented:
Yay!

dtzWill commented on pkgs/os-specific/linux/mmc-utils/default.nix:
```diff
@@ -10,6 +10,8 @@ stdenv.mkDerivation rec {
     sha256 = "0hkdzc71pdnscbpdpgwljcchiyancarldjyd0w609sy18bky833x";
   };
 
+  makeFlags = "CC=${stdenv.cc.targetPrefix}cc";
```
I don't believe this is needed, here's cross-building to aarch64-musl:

```console
$ nix build -f channel:nixos-unstable --arg crossSystem '(import <nixpkgs/lib>).systems.examples.aarch64-multiplatform-musl' mmc-utils
$ ls -l result
lrwxrwxrwx 1 will users 91 May  3 16:34 result -> /nix/store/g5izr5yd8sydw6dizmnj8y2mlvwx28a5-mmc-utils-2015-11-18-aarch64-unknown-linux-musl
```

(where unstable is based on master, and doesn't have this commit yet)

bgamari replied:
The same here.

dtzWill commented on pkgs/os-specific/linux/net-tools/default.nix:
```diff
@@ -15,6 +15,8 @@ stdenv.mkDerivation rec {
 
   makeFlags = [
+    "CC=${stdenv.cc.targetPrefix}cc"
+    "AR=${stdenv.cc.targetPrefix}ar"
```
These don't appear to be needed either. Same command as above, but try building nettools.

bgamari replied:
Hmm, I could have sworn I cherry-picked this one for a reason but indeed I haven't seen a failure yet after reverting it. Very odd.

dtzWill commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -76,8 +76,8 @@ stdenv.mkDerivation rec {
-  CXX = "c++";
-  CXXCPP = "c++ -E";
+  CXX = "${stdenv.cc.targetPrefix}c++";
+  CXXCPP = "${stdenv.cc.targetPrefix}c++ -E";
```
This works as-is for me as well?

```console
$ nix build -f channel:nixos-unstable --arg crossSystem '(import <nixpkgs/lib>).systems.examples.aarch64-multiplatform-musl' curl
```

bgamari replied:
I'll admit I didn't test this one. I had assumed that `c++` would either not exist or refer to the host compiler.

dtzWill commented on pkgs/tools/security/rng-tools/default.nix:
```diff
@@ -9,6 +9,9 @@ stdenv.mkDerivation rec {
 
+  # For cross-compilation
+  makeFlags = [ "AR:=$(AR)" ];
+
```
Yes, this is needed! :grin:

```console
$ nix build -f channel:nixos-unstable --arg crossSystem '(import <nixpkgs/lib>).systems.examples.aarch64-multiplatform-musl' rng_tools                            16:40:32 on 18-05-03
error: build of '/nix/store/4sh7pnz0vwacxm7a784fyf6q6cbzjhaw-rng-tools-5-aarch64-unknown-linux-musl.drv' on 'ssh://wdietz2@presto' failed: builder for '/nix/store/4sh7pnz0vwacxm7a784fyf6q6cbzjhaw-rng-tools-5-aarch64-unknown-linux-musl.drv' failed with exit code 2
builder for '/nix/store/4sh7pnz0vwacxm7a784fyf6q6cbzjhaw-rng-tools-5-aarch64-unknown-linux-musl.drv' failed with exit code 1; last 10 log lines:
  aarch64-unknown-linux-musl-gcc -DHAVE_CONFIG_H -I.     -g -O2 -c fips.c
  rm -f librngd.a
  ar cru librngd.a fips.o
  /nix/store/xn5gv3lpfy91yvfy9b0i7klfcxh9xskz-bash-4.4-p19/bin/bash: ar: command not found
  make[2]: *** [Makefile:440: librngd.a] Error 127
  make[2]: Leaving directory '/build/rng-tools-5'
  make[1]: *** [Makefile:672: all-recursive] Error 1
  make[1]: Leaving directory '/build/rng-tools-5'
  make: *** [Makefile:377: all] Error 2
  builder for '/nix/store/4sh7pnz0vwacxm7a784fyf6q6cbzjhaw-rng-tools-5-aarch64-unknown-linux-musl.drv' failed with exit code 2
[0 built (1 failed), 0.0 MiB DL]
error: build of '/nix/store/4sh7pnz0vwacxm7a784fyf6q6cbzjhaw-rng-tools-5-aarch64-unknown-linux-musl.drv' failed
```

dtzWill reviewed (commented):
These don't all seem to be needed--see comments.

If they're needed in your environment, I'd be curious to sort out why that is and for improving my own testing :).

Ericson2314 commented:
@dtzWill Thanks for investigating, and my bad on the eager merge. I guess just revert those bits before staging evals again?

bgamari commented:
Sorry for the confusion everyone! I could have sworn those two were needed.
